src/librbd/ImageWatcher.cc

```diff
@@ -393,37 +393,37 @@ void ImageWatcher<I>::notify_quiesce(uint64_t *request_id,
 
   AsyncRequestId async_request_id(get_client_id(), *request_id);
 
-  auto attempts = m_image_ctx.config.template get_val<uint64_t>(
+  auto total_attempts = m_image_ctx.config.template get_val<uint64_t>(
     "rbd_quiesce_notification_attempts");
 
-  notify_quiesce(async_request_id, attempts, prog_ctx, on_finish);
+  notify_quiesce(async_request_id, 1, total_attempts, prog_ctx, on_finish);
 }
 
 template <typename I>
 void ImageWatcher<I>::notify_quiesce(const AsyncRequestId &async_request_id,
-                                     size_t attempts, ProgressContext &prog_ctx,
+                                     size_t attempt, size_t total_attempts,
+                                     ProgressContext &prog_ctx,
                                      Context *on_finish) {
+  ceph_assert(total_attempts > 0);
   ldout(m_image_ctx.cct, 10) << this << " " << __func__ << ": async_request_id="
-                             << async_request_id << " attempts=" << attempts
-                             << dendl;
+                             << async_request_id << " attempts @ "
+                             << attempt << "/" << total_attempts << dendl;
 
-  ceph_assert(attempts > 0);
   auto notify_response = new watcher::NotifyResponse();
   auto on_notify = new LambdaContext(
     [notify_response=std::unique_ptr<watcher::NotifyResponse>(notify_response),
-     this, async_request_id, &prog_ctx, on_finish, attempts=attempts-1](int r) {
-      auto total_attempts = m_image_ctx.config.template get_val<uint64_t>(
-        "rbd_quiesce_notification_attempts");
-      if (total_attempts < attempts) {
-        total_attempts = attempts;
-      }
-      prog_ctx.update_progress(total_attempts - attempts, total_attempts);
-
+     this, async_request_id, &prog_ctx, on_finish, attempt=attempt+1,
+     total_attempts](int r) {
+      prog_ctx.update_progress(attempt-1, total_attempts);
       if (r == -ETIMEDOUT) {
-        ldout(m_image_ctx.cct, 10) << this << " " << __func__ << ": async_request_id="
-                                   << async_request_id << " timed out" << dendl;
-        if (attempts > 0) {
-          notify_quiesce(async_request_id, attempts, prog_ctx, on_finish);
+        ldout(m_image_ctx.cct, 10) << this << " " << __func__
+                                   << ": async_request_id=" << async_request_id
+                                   << " timed out" << dendl;
+        auto total_attempts = m_image_ctx.config.template get_val<uint64_t>(
+          "rbd_quiesce_notification_attempts");
```

Comment:
@idryomov @trociny,

Say, the initial total_attempts = 10 and the current attempt is 6 i.e. 6/10.
Assuming the user reduced the `rbd_quiesce_notification_attempts` to 5. Then the condition below fails and notify_quiesce() will not be called. This lead to confusion in the logs about why the notify_quiesce() stopped a 6/10 even though the previous request is `timed out`.

To avoid this case is it worth log `rbd_quiesce_notification_attempts` on get_val ? (both at line 396 & 422?)

Reply:
I don't think we need to consult `rbd_quiesce_notification_attempts` anywhere other than before the initial attempt.  This potential confusion becomes moot then.

Reply:
Your `static total_attempts` version would have captured `rbd_quiesce_notification_attempts` value for the lifetime of the process, which in case of a long-running VM can be months if not years.  With my suggestion, it's captured before each "snap create" operation but stays the same between quiesce notification attempts within the "snap create" operation to keep things simple and avoid confusion of the kind that you brought up in this thread.